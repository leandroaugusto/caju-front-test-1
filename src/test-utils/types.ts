export interface ITestUtils {
  children?: React.ReactNode
}
