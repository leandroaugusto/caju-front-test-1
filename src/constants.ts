const isTestEnv = process.env.NODE_ENV === 'test';

export const serverHost: string = isTestEnv
  ? process.env.VITE_SERVER_HOST
  : import.meta.env.VITE_SERVER_HOST;

export const queryKey: string = isTestEnv
  ? process.env.VITE_QUERY_KEY
  : import.meta.env.VITE_QUERY_KEY;
