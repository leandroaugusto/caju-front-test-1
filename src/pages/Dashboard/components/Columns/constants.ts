import { ERegistrationsStatus } from "~/types/registrations.types";

export const allColumns = [
  { status: ERegistrationsStatus.REVIEW, title: "Pronto para revisar" },
  { status: ERegistrationsStatus.APPROVED, title: "Aprovado" },
  { status: ERegistrationsStatus.REPROVED, title: "Reprovado" },
];
