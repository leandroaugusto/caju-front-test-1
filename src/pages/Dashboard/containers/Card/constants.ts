export const promptMessages = {
  approve: "Deseja aprovar a admiss達o de",
  reprove: "Deseja reprovar a admiss達o de",
  review: "Deseja revisar a admiss達o de",
  delete: "Deseja excluir a admiss達o de",
}
