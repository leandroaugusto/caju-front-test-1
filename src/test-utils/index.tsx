import { ReactElement } from "react";
import { ThemeProvider } from "styled-components";
import { render, RenderOptions, RenderResult } from "@testing-library/react";

import { ITestUtils } from "./types";

import theme from "~/theme";

const AllTheProviders: React.FC = ({ children }: ITestUtils) => {
  return <ThemeProvider theme={theme}>{children}</ThemeProvider>;
};

const customRender = (
  ui: ReactElement,
  options?: Omit<RenderOptions, "wrapper">
): RenderResult =>
  render(ui, {
    wrapper: AllTheProviders,
    ...options,
  });

export * from "@testing-library/react";

export { customRender };
