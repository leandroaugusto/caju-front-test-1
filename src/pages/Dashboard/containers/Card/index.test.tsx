import { faker } from "@faker-js/faker";

import { customRender, fireEvent } from "~/test-utils";
import {
  TRegistrationsData,
  ERegistrationsStatus,
} from "~/types/registrations.types";

import type { ISnackbarProps } from "~/components/Snackbar/types";
import type { IModalProps } from "~/components/Modal/types";
import type { IRegistrationCardProps } from "../../components/RegistrationCard/types";

import { Card } from ".";
import { promptMessages } from "./constants";

const reviewUserMutation = vi.fn();
const deleteUserMutation = vi.fn();
const deleteUser = {
  mutateAsync: deleteUserMutation,
};
const reviewUser = {
  mutateAsync: reviewUserMutation,
};

const registrationMock: TRegistrationsData = {
  admissionDate: "22/10/2023",
  email: "luiz@caju.com.br",
  employeeName: "Luiz Filho",
  status: ERegistrationsStatus.REVIEW,
  cpf: faker.string.numeric(11),
  id: faker.string.numeric(),
};

const mockedLocation = vi.hoisted(() =>
  vi.fn(() => ({ state: "Operação concluída com sucesso" }))
);

vi.mock("react-router-dom", () => ({
  useLocation: mockedLocation,
}));

vi.mock("~/hooks/useRegistrations", () => ({
  useRegistrationsHook: () => ({
    deleteUser,
    reviewUser,
  }),
}));

vi.mock("~/components/Snackbar", () => ({
  SnackBar: ({ message }: ISnackbarProps) => (
    <div data-testid="mock-snackbar">{message}</div>
  ),
}));

vi.mock("../../components/RegistrationCard", () => ({
  RegistrationCard: ({ onCardAction }: IRegistrationCardProps) => (
    <div data-testid="mock-registration-card">
      <button
        data-testid="mock-card-action-delete"
        onClick={() => onCardAction("delete")}
      >
        Delete
      </button>
      <button
        data-testid="mock-card-action-approve"
        onClick={() => onCardAction("approve")}
      >
        Approve
      </button>
      <button
        data-testid="mock-card-action-reprove"
        onClick={() => onCardAction("reprove")}
      >
        Reprove
      </button>
      <button
        data-testid="mock-card-action-review"
        onClick={() => onCardAction("review")}
      >
        Review
      </button>
    </div>
  ),
}));

vi.mock("~/components/Modal", () => ({
  Modal: ({ onConfirm, message }: IModalProps) => (
    <div data-testid="mock-modal">
      <span data-testid="mock-modal-message">{message}</span>
      <button data-testid="mock-modal-confirm" onClick={onConfirm}>
        onConfirm
      </button>
    </div>
  ),
}));

vi.mock("~/components/Loading", () => ({
  Loading: () => <div data-testid="mock-loading" />,
}));

describe("Card", () => {
  it("Should show Card container", () => {
    const { container } = customRender(<Card data={registrationMock} />);
    expect(container).toMatchSnapshot();
  });

  it("should display error message in snackbar when async operation fails", async () => {
    const errorMessage = "Error occurred";
    reviewUserMutation.mockRejectedValue(new Error(errorMessage));

    const { getByTestId, findByText } = customRender(
      <Card data={registrationMock} />
    );

    const approveButton = getByTestId("mock-card-action-approve");
    fireEvent.click(approveButton);

    const confirmButton = getByTestId("mock-modal-confirm");
    fireEvent.click(confirmButton);

    expect(await findByText(errorMessage)).toBeInTheDocument();
  });

  it("should display success message in snackbar when async operation succeeds", async () => {
    const modalMessage = `${promptMessages.approve} ${registrationMock.employeeName}?`;
    const { getByTestId } = customRender(<Card data={registrationMock} />);

    const approveButton = getByTestId("mock-card-action-approve");
    fireEvent.click(approveButton);

    const confirmButton = getByTestId("mock-modal-confirm");
    const modalMessageContainer = getByTestId("mock-modal-message");

    expect(modalMessageContainer).toHaveTextContent(modalMessage);

    fireEvent.click(confirmButton);

    expect(mockedLocation).toHaveBeenCalledTimes(3);
  });

  it("should reprove a registration", async () => {
    const { getByTestId } = customRender(<Card data={registrationMock} />);

    const reproveButton = getByTestId("mock-card-action-reprove");
    fireEvent.click(reproveButton);

    const confirmButton = getByTestId("mock-modal-confirm");
    fireEvent.click(confirmButton);

    expect(mockedLocation).toHaveBeenCalledTimes(3);
  });

  it("should review a registration", async () => {
    registrationMock.status = ERegistrationsStatus.APPROVED;
    const { getByTestId } = customRender(<Card data={registrationMock} />);

    const reviewButton = getByTestId("mock-card-action-review");
    fireEvent.click(reviewButton);

    const confirmButton = getByTestId("mock-modal-confirm");
    fireEvent.click(confirmButton);

    expect(mockedLocation).toHaveBeenCalledTimes(3);
  });

  it("should delete a registration", async () => {
    const { getByTestId } = customRender(<Card data={registrationMock} />);

    const deleteButton = getByTestId("mock-card-action-delete");
    fireEvent.click(deleteButton);

    const confirmButton = getByTestId("mock-modal-confirm");
    fireEvent.click(confirmButton);

    expect(mockedLocation).toHaveBeenCalledTimes(3);
  });
});
