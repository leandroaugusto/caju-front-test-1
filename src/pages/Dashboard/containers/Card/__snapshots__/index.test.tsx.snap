// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Card > Should show Card container 1`] = `
<div>
  <div
    data-testid="mock-registration-card"
  >
    <button
      data-testid="mock-card-action-delete"
    >
      Delete
    </button>
    <button
      data-testid="mock-card-action-approve"
    >
      Approve
    </button>
    <button
      data-testid="mock-card-action-reprove"
    >
      Reprove
    </button>
    <button
      data-testid="mock-card-action-review"
    >
      Review
    </button>
  </div>
  <div
    data-testid="mock-snackbar"
  />
  <div
    data-testid="mock-modal"
  >
    <span
      data-testid="mock-modal-message"
    >
      Deseja revisar a admissão de Luiz Filho?
    </span>
    <button
      data-testid="mock-modal-confirm"
    >
      onConfirm
    </button>
  </div>
</div>
`;

exports[`Card > Should show Card with another status 1`] = `
<div>
  <div
    data-testid="mock-registration-card"
  >
    <button
      data-testid="mock-card-action-delete"
    >
      Delete
    </button>
    <button
      data-testid="mock-card-action-approve"
    >
      Approve
    </button>
    <button
      data-testid="mock-card-action-reprove"
    >
      Reprove
    </button>
    <button
      data-testid="mock-card-action-review"
    >
      Review
    </button>
  </div>
  <div
    data-testid="mock-snackbar"
  />
  <div
    data-testid="mock-modal"
  >
    <span
      data-testid="mock-modal-message"
    >
      Deseja revisar a admissão de Luiz Filho?
    </span>
    <button
      data-testid="mock-modal-confirm"
    >
      onConfirm
    </button>
  </div>
</div>
`;
